### PR TITLE
issue #97: Import externallib with use statements

### LINKED ISSUE #97: PHP Fatal error runinseparate process
Moodle MOODLE_402_STABLE
elastic MOODLE_310_STABLE
PHP8.0

```
PHP Fatal error:  Uncaught coding_exception: Coding error detected, it must be fixed by a programmer: When including this file for a unit test, the test must be run in an isolated process. See the PHPUnit @runInSeparateProcess and @runTestsInSeparateProcesses annotations. in /var/www/html/lib/setuplib.php:2204
    Stack trace:
    #0 /var/www/html/lib/externallib.php(35): require_phpunit_isolation()
    #1 /var/www/html/search/engine/elastic/externallib.php(26): require_once('/var/www/html/l...')
    #2 /var/www/html/search/engine/elastic/tests/externallib_test.php(33): require_once('/var/www/html/s...')
    #3 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(66): include_once('/var/www/html/s...')
    #4 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(49): PHPUnit\Util\FileLoader::load('/var/www/html/s...')
    #5 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(397): PHPUnit\Util\FileLoader::checkAndLoad('/var/www/html/s...')
    #6 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(527): PHPUnit\Framework\TestSuite->addTestFile('/var/www/html/s...')
    #7 /var/www/html/vendor/phpunit/phpunit/src/TextUI/TestSuiteMapper.php(67): PHPUnit\Framework\TestSuite->addTestFiles(Array)
    #8 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(391): PHPUnit\TextUI\TestSuiteMapper->map(Object(PHPUnit\TextUI\XmlConfiguration\TestSuiteCollection), '')
    #9 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(112): PHPUnit\TextUI\Command->handleArguments(Array)
    #10 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(97): PHPUnit\TextUI\Command->run(Array, true)
    #11 /var/www/html/vendor/phpunit/phpunit/phpunit(98): PHPUnit\TextUI\Command::main()
    #12 /var/www/html/vendor/bin/phpunit(122): include('/var/www/html/v...')
    #13 {main}
  
    Next PHPUnit\TextUI\RuntimeException: Coding error detected, it must be fixed by a programmer: When including this file for a unit test, the test must be run in an isolated process. See the PHPUnit @runInSeparateProcess and @runTestsInSeparateProcesses annotations. in /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php:99
    Stack trace:
    #0 /var/www/html/vendor/phpunit/phpunit/phpunit(98): PHPUnit\TextUI\Command::main()
    #1 /var/www/html/vendor/bin/phpunit(122): include('/var/www/html/v...')
    #2 {main}
      thrown in /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php on line 99
  stderr_lines: <omitted>
  stdout: |-
    Compose v2 is not available in Docker CLI, falling back to use docker-compose script
    Moodle 4.2+ (Build: 20230526), 6c9d8e17ddcdceb4bcb0c7db23b39ff3b01c624a
    Php: 8.0.28, mysqli: 8.0.33, OS: Linux 5.15.73-45.135.amzn2022.x86_64 x86_64
  
    Fatal error: Uncaught coding_exception: Coding error detected, it must be fixed by a programmer: When including this file for a unit test, the test must be run in an isolated process. See the PHPUnit @runInSeparateProcess and @runTestsInSeparateProcesses annotations. in /var/www/html/lib/setuplib.php:2204
    Stack trace:
    #0 /var/www/html/lib/externallib.php(35): require_phpunit_isolation()
    #1 /var/www/html/search/engine/elastic/externallib.php(26): require_once('/var/www/html/l...')
    #2 /var/www/html/search/engine/elastic/tests/externallib_test.php(33): require_once('/var/www/html/s...')
    #3 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(66): include_once('/var/www/html/s...')
    #4 /var/www/html/vendor/phpunit/phpunit/src/Util/FileLoader.php(49): PHPUnit\Util\FileLoader::load('/var/www/html/s...')
    #5 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(397): PHPUnit\Util\FileLoader::checkAndLoad('/var/www/html/s...')
    #6 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestSuite.php(527): PHPUnit\Framework\TestSuite->addTestFile('/var/www/html/s...')
    #7 /var/www/html/vendor/phpunit/phpunit/src/TextUI/TestSuiteMapper.php(67): PHPUnit\Framework\TestSuite->addTestFiles(Array)
    #8 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(391): PHPUnit\TextUI\TestSuiteMapper->map(Object(PHPUnit\TextUI\XmlConfiguration\TestSuiteCollection), '')
    #9 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(112): PHPUnit\TextUI\Command->handleArguments(Array)
    #10 /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php(97): PHPUnit\TextUI\Command->run(Array, true)
    #11 /var/www/html/vendor/phpunit/phpunit/phpunit(98): PHPUnit\TextUI\Command::main()
    #12 /var/www/html/vendor/bin/phpunit(122): include('/var/www/html/v...')
    #13 {main}
  
    Next PHPUnit\TextUI\RuntimeException: Coding error detected, it must be fixed by a programmer: When including this file for a unit test, the test must be run in an isolated process. See the PHPUnit @runInSeparateProcess and @runTestsInSeparateProcesses annotations. in /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php:99
    Stack trace:
    #0 /var/www/html/vendor/phpunit/phpunit/phpunit(98): PHPUnit\TextUI\Command::main()
    #1 /var/www/html/vendor/bin/phpunit(122): include('/var/www/html/v...')
    #2 {main}
      thrown in /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php on line 99
```

### PR DESCRIPTION
**Pull request by ScottVerbeek**
Resolves coding_exception in MOODLE_402_STABLE and up.

These tests were executed on a 4.3+ branch:
```
vendor/bin/phpunit --testsuite=search_elastic_testsuite
Moodle 4.3+ (Build: 20231027), e57253711b64132738b2496e6bbfc8ebb02bbb3d
Php: 8.1.24, pgsql: 14.0 (Debian 14.0-1.pgdg110+1), OS: Linux 6.4.12-060412-generic x86_64
PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

...............................................................  63 / 110 ( 57%)
...............................................                 110 / 110 (100%)

Time: 01:13.444, Memory: 95.50 MB

OK (110 tests, 217 assertions)
```

Closes  #97

### CODE REVIEW

**Comment by ScottVerbeek** [comment]
The tests performed earlier on this unit test we not on the version of Moodle 4.3. I have now updated locally and amended the commit resolving one of two problems. I'm still looking into this failing unit test.
```
vendor/bin/phpunit --testsuite search_elastic_testsuite
Moodle 4.3.2 (Build: 20231222), f810374a0ec816eca849f807cb09ff0c478438b7
Php: 8.1.24, pgsql: 14.0 (Debian 14.0-1.pgdg110+1), OS: Linux 6.4.12-060412-generic x86_64
PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

.....................................F.........................  63 / 115 ( 54%)
....................................................            115 / 115 (100%)

Time: 01:22.411, Memory: 80.50 MB

There was 1 failure:

1) search_elastic\esrequest_test::test_get_unreachable_host
Failed asserting that exception of type "GuzzleHttp\Exception\ConnectException" is thrown.

/var/www/vanilla-403/lib/phpunit/classes/advanced_testcase.php:81

FAILURES!
Tests: 115, Assertions: 223, Failures: 1.
```

**Comment by ScottVerbeek** [comment]
> ```
> There was 1 failure:
> 
> 1) search_elastic\esrequest_test::test_get_unreachable_host
> Failed asserting that exception of type "GuzzleHttp\Exception\ConnectException" is thrown.
> 
> /var/www/vanilla-403/lib/phpunit/classes/advanced_testcase.php:81
> ```

This failure has been relation to my local environment. I have setup a simpler one and the tests are now all passing. This MR is ready for a review.

**Comment by ScottVerbeek** [comment]
Due to https://tracker.moodle.org/browse/MDL-76583 this will need a separate branch. I'll create MOODLE_402_STABLE and update the readme.

**Comment by ScottVerbeek** [comment]
Once this has been mered, I'll create merge requests into the other branches updating the readme.  Once those have been merged I'll make the branch `MOODLE_402_STABLE` the new default branch.

File: .github/workflows/moodle-release.yml [hunk 1]
```diff
@@ -2,18 +2,18 @@
 # Whenever version.php is changed, add the latest version
 # to the Moodle Plugins directory at https://moodle.org/plugins
 #
-name: MOODLE_310_STABLE - Releasing in the Plugins directory
```
Comment (keevan) [inline .github/workflows/moodle-release.yml:5]: Might be worth combining the release logic into ci.yml (per the latest reusable workflows docs)

File: README.md [hunk 2]
```diff
@@ -17,6 +17,7 @@ This plugin currently supports Moodle:
 
 | Moodle version      | Branch               |
 | ------------------- | -------------------- |
+| Moodle 4.2 and up   | MOODLE_402_STABLE    |
```
Comment (keevan) [inline README.md:20]: Are 4.0 and 4.1 not supported, if they are, which branch should be used? Seems to be a gap here. 

Could you check this please.

**Review by keevan — COMMENTED** [review]
Functional changes here look okay. 

In addition to the notes, are you able to check the CI, there seems to be some errors.

File: README.md (hunk 2 above)
Comment (ScottVerbeek) [reply README.md:20]: Due to https://tracker.moodle.org/browse/MDL-76583 Moodle 4.0 and 4.1 are not supported. I have amended the readme commit accordingly. Even though the commit introduces the namespace I have also given this a test locally. Here are the results of that.

<details>
<summary>PHPUnit tests results with changes on Moodle 4.1.8 (Build: 20231222)</summary>

```
root@b0ba46b747f2:/var/www/site# vendor/bin/phpunit --testsuite search_elastic_testsuite
Moodle 4.1.8 (Build: 20231222), e591ddc4e8c21df9fb5369e388ef57ef07acbb9b
Php: 8.1.24, pgsql: 14.0 (Debian 14.0-1.pgdg110+1), OS: Linux 6.4.12-060412-generic x86_64
PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

...............................................EE..............  63 / 115 ( 54%)
....................................................            115 / 115 (100%)

Time: 02:21.574, Memory: 78.50 MB

There were 2 errors:

1) search_elastic\externallib_test::test_external_search
Error: Class "core_external\external_api" not found

/var/www/site/search/engine/elastic/externallib.php:40
/var/www/site/search/engine/elastic/tests/externallib_test.php:103
/var/www/site/lib/phpunit/classes/advanced_testcase.php:80

2) search_elastic\externallib_test::test_external_search_areas
Error: Class "search_elastic_external" not found

/var/www/site/search/engine/elastic/tests/externallib_test.php:172
/var/www/site/lib/phpunit/classes/advanced_testcase.php:80

ERRORS!
Tests: 115, Assertions: 219, Errors: 2.
```
</details>

**Review by keevan — APPROVED** [review]
Looks good, thanks Scott.

**Pull request merged**